service-base/src/test/java/org/eclipse/hono/service/management/device/AutoProvisioningEnabledDeviceBackendTest.java

```diff
@@ -0,0 +1,95 @@
+import java.security.cert.CertificateFactory;
+import java.security.cert.X509Certificate;
+import java.util.Optional;
+
+import org.eclipse.hono.service.management.Id;
+import org.eclipse.hono.service.management.OperationResult;
+import org.junit.jupiter.api.BeforeAll;
+import org.junit.jupiter.api.Test;
+import org.junit.jupiter.api.extension.ExtendWith;
+
+import io.opentracing.noop.NoopSpan;
+import io.vertx.core.Future;
+import io.vertx.core.Promise;
+import io.vertx.core.net.SelfSignedCertificate;
+import io.vertx.junit5.VertxExtension;
+import io.vertx.junit5.VertxTestContext;
+
+/**
+ * Tests verifying behavior of {@link AutoProvisioningEnabledDeviceBackend}.
+ */
+@ExtendWith(VertxExtension.class)
+class AutoProvisioningEnabledDeviceBackendTest {
+
+    private static final String TENANT_ID = "test-tenant";
+    private static final String DEVICE_ID = "4711";
+
+    private static X509Certificate cert;
+
+    @BeforeAll
+    static void before() throws GeneralSecurityException, IOException {
+        final SelfSignedCertificate ssc = SelfSignedCertificate.create("eclipse.org");
+        final CertificateFactory factory = CertificateFactory.getInstance("X.509");
+        cert = (X509Certificate) factory.generateCertificate(new FileInputStream(ssc.certificatePath()));
+    }
+
+    /**
+     * Verifies that a device is created and credentials are set for it when auto-provisioning a device.
+     */
+    @Test
+    void provisionDevice(final VertxTestContext ctx) {
+
+        // GIVEN an AutoProvisioningEnabledDeviceBackend instance with "happy path" answers
+        final AutoProvisioningEnabledDeviceBackend underTest = mock(AutoProvisioningEnabledDeviceBackend.class);
+        when(underTest.provisionDevice(anyString(), any(), any())).thenCallRealMethod();
+
+        doAnswer(invocation -> {
+            final Promise<OperationResult<Id>> promise = invocation.getArgument(4);
+            promise.complete(OperationResult.ok(201, Id.of(DEVICE_ID), Optional.empty(), Optional.empty()));
+            return null;
+        }).when(underTest).createDevice(any(), any(), any(), any(), any(Promise.class));
+
+        doAnswer(invocation -> {
+            final Promise<OperationResult<Void>> promise = invocation.getArgument(5);
+            promise.complete(OperationResult.empty(204));
+            return null;
+        }).when(underTest).set(any(), any(), any(), any(), any(), any(Promise.class));
+
+        // WHEN provisioning a device from a certificate
+        final Future<OperationResult<String>> result = underTest.provisionDevice(TENANT_ID, cert, NoopSpan.INSTANCE);
+
+        // THEN the device is created and credentials are set
+        result.setHandler(ctx.succeeding(ok -> {
+            verify(underTest).createDevice(eq(TENANT_ID), any(), any(), any(), any(Promise.class));
+            verify(underTest).set(eq(TENANT_ID), eq(DEVICE_ID), any(), any(), any(), any(Promise.class));
+            ctx.completeNow();
+        }));
+    }
+
```

Comment:
what about a test for verifying that the device identity is removed if creating the credentials fails?